From 93e31c178f9781ba841e80e8568461c0b99c1b88 Mon Sep 17 00:00:00 2001
From: David Heidelberg <david@ixit.cz>
Date: Thu, 15 Aug 2024 03:14:16 -0500
Subject: [PATCH 094/211] arm64: dts: qcom: sdm845-xiaomi-beryllium: Add
 support for imx363 camera

Wire the Sony IMX363 camera.

Based on work of Joel Selvaraj.

Signed-off-by: David Heidelberg <david@ixit.cz>
---
 .../qcom/sdm845-xiaomi-beryllium-common.dtsi  | 107 +++++++++++++++++-
 1 file changed, 106 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/qcom/sdm845-xiaomi-beryllium-common.dtsi b/arch/arm64/boot/dts/qcom/sdm845-xiaomi-beryllium-common.dtsi
index 40fbca97a1bc..9a877e76b908 100644
--- a/arch/arm64/boot/dts/qcom/sdm845-xiaomi-beryllium-common.dtsi
+++ b/arch/arm64/boot/dts/qcom/sdm845-xiaomi-beryllium-common.dtsi
@@ -151,6 +151,42 @@ vreg_s4a_1p8: vreg-s4a-1p8 {
 		regulator-max-microvolt = <1800000>;
 		regulator-always-on;
 	};
+
+	cam_iovdd: reg_cam_iovdd {
+		compatible = "regulator-fixed";
+		regulator-name = "cam_iovdd";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <1800000>;
+		enable-active-high;
+		gpio = <&tlmm 50 GPIO_ACTIVE_HIGH>;
+	};
+
+	cam_ldo: reg_cam_ldo {
+		compatible = "regulator-fixed";
+		regulator-name = "cam_ldo";
+		regulator-min-microvolt = <1350000>;
+		regulator-max-microvolt = <1350000>;
+		enable-active-high;
+		gpio = <&tlmm 99 GPIO_ACTIVE_HIGH>;
+		regulator-always-on; // FIXME
+	};
+
+	cam0_avdd: reg_cam0_avdd {
+		compatible = "regulator-fixed";
+		regulator-name = "cam0_avdd";
+		enable-active-high; // FIXME
+		gpio = <&tlmm 40 GPIO_ACTIVE_HIGH>;
+	};
+
+	cam0_dvdd: reg_cam0_dvdd {
+		compatible = "regulator-fixed";
+		regulator-name = "cam0_dvdd";
+		regulator-min-microvolt = <1350000>;
+		regulator-max-microvolt = <1350000>;
+		regulator-enable-ramp-delay = <135>;
+		enable-active-high; // FIXME
+		gpio = <&pm8998_gpios 11 GPIO_ACTIVE_HIGH>;
+	};
 };
 
 &adsp_pas {
@@ -247,10 +283,61 @@ vreg_l26a_1p2: ldo26 {
 	};
 };
 
+&cci {
+	status = "okay";
+};
+
+&camss {
+	vdda-phy-supply = <&vreg_l1a_0p875>;
+	vdda-pll-supply = <&vreg_l26a_1p2>;
+
+	status = "okay";
+
+	ports {
+		#address-cells = <1>;
+		#size-cells = <0>;
+		port@0 {
+			reg = <0>;
+			csiphy0_ep: endpoint {
+				clock-lanes = <7>;
+				data-lanes = <0 1 2 3>;
+				remote-endpoint = <&imx363_ep>;
+			};
+		};
+	};
+};
+
 &cci_i2c0 {
 	status = "okay";
 
-	/* IMX363 @ 10 */
+	camera@10 {
+		compatible = "sony,imx363";
+		reg = <0x10>;
+
+		reset-gpios = <&tlmm 80 GPIO_ACTIVE_LOW>;
+		pinctrl-0 = <&cam0_default &cam_mclk0_default>;
+		pinctrl-names = "default";
+
+		clocks = <&clock_camcc CAM_CC_MCLK0_CLK>;
+		clock-names = "xvclk";
+		clock-frequency = <24000000>;
+
+		vif-supply = <&cam_iovdd>;
+		vana-supply = <&cam0_avdd>;
+		vdig-supply = <&cam0_dvdd>;
+
+		orientation = <1>;
+		rotation = <270>;
+
+		port {
+			imx363_ep: endpoint {
+				clock-lanes = <1>;
+				data-lanes = <1 2 3 4>;
+				link-frequencies = /bits/ 64 <636000000>;
+				remote-endpoint = <&csiphy0_ep>;
+			};
+		};
+	};
 };
 
 &cdsp_pas {
@@ -606,6 +693,24 @@ codec {
 &tlmm {
 	gpio-reserved-ranges = <0 4>, <81 4>;
 
+	cam0_default: cam-rear-default-state {
+		rst-pins {
+			pins = "gpio80";
+			function = "gpio";
+
+			drive-strength = <2>;
+			bias-disable;
+		};
+
+		avdd-pins {
+			pins = "gpio40";
+			function = "gpio";
+
+			drive-strength = <2>;
+			bias-disable;
+		};
+	};
+
 	sdc2_default_state: sdc2-default-state {
 		clk-pins {
 			pins = "sdc2_clk";
-- 
2.55.0

